### PR TITLE
Add forwarded-header-policy enhancement

### PR DESCRIPTION
**Pull request by Miciah**

### CODE REVIEW

File: enhancements/ingress/forwarded-header-policy.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+HAProxy sets the `X-Forwarded-For` header, as well as `http-request add-header`
+and `http-request set-header` keywords to append or replace arbitrary headers.
+The `option forwardfor` keyword has an associated `if-none` keyword, which can
+be used to implement the homonymous IngressController policy.  The `http-request
+set-header` keyword can be conditionalized using ACLs to implement `if-none`.
+These keywords are sufficient to implement all four policies.
+
+The HAProxy configuration template generates a backend configuration block for
+each route based on the value in the route's annotation or the value of
+`ROUTER_SET_FORWARDED_HEADERS`.  To implement `append` for the `X-Forwarded-For`
```
Comment (sgreene570) [inline enhancements/ingress/forwarded-header-policy.md:258]: > based on the value in the route's annotation or the value of
`ROUTER_SET_FORWARDED_HEADERS`.

If the route's annotation and `ROUTER_SET_FORWARDED_HEADERS` specify a forwarded header policy, the route's annotation will override `ROUTER_SET_FORWARDED_HEADERS`, correct?

File: enhancements/ingress/forwarded-header-policy.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+Omitting `spec.httpHeaders` or omitting `spec.httpHeaders.forwarded` specifies
+the default behavior.
+
+The API validates the `spec.httpHeaders.forwarded` field value as described by
+the field type's `+kubebuilder:validation:Enum` marker.
+
+The HAProxy configuration template validates the
+`haproxy.router.openshift.io/set-forwarded-headers` annotation.  If the
+annotation specifies an invalid value, then the IngressController uses its
+configured policy.
```
Comment (danehans) [inline enhancements/ingress/forwarded-header-policy.md:208]: > If the annotation specifies an invalid value, then the IngressController uses its configured policy.

Maybe we should reflect the policy used by the ingress controller in status.

File: enhancements/ingress/forwarded-header-policy.md (hunk 1 above)
Comment (Miciah) [reply enhancements/ingress/forwarded-header-policy.md:258]: Correct.

File: enhancements/ingress/forwarded-header-policy.md (hunk 2 above)
Comment (Miciah) [reply enhancements/ingress/forwarded-header-policy.md:208]: Do you mean reflect the policy in the ingress controller's status, or in the route's?

**Comment by Miciah** [comment]
Latest push changes the status to "implemented", updates the test plan and documentation status, updates the API definitions to match what was merged in openshift/api, and fixes some minor errors of spelling and grammar.

File: enhancements/ingress/forwarded-header-policy.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+	// * "IfNone", which specifies that the IngressController sets the
+	//   headers if they are not already set.
+	//
+	// * "Never", which specifies that the IngressController never sets the
+	//   headers, preserving any existing headers.
+	//
+	// By default, the policy is "Append".
+	//
+	// +optional
+	ForwardedHeaderPolicy IngressControllerHTTPHeaderPolicy `json:"forwardedHeaderPolicy,omitempty"`
```
Comment (sttts) [inline enhancements/ingress/forwarded-header-policy.md:138]: set the default as CRD marker

File: enhancements/ingress/forwarded-header-policy.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+spec:
+  to:
+    kind: Service
+    name: hello-openshift
+```
+
+#### Validation
+
+Omitting `spec.httpHeaders` or omitting `spec.httpHeaders.forwardedHeaderPolicy`
+specifies the default behavior.
```
Comment (sttts) [inline enhancements/ingress/forwarded-header-policy.md:184]: be precize. Is this defaulted in the CRD or is this just default behaviour of the controllers?

File: enhancements/ingress/forwarded-header-policy.md (hunk 3 above)
Comment (sttts) [reply enhancements/ingress/forwarded-header-policy.md:138]: There is no risk.

**Comment by sttts** [comment]
Sgtm. Some team member has to lgtm.

**Pull request closed without merging**